File: domain/classification_level.py
"""
domain/classification_level.py

ClassificationLevel domain model.
Mirrors contracts/common/common-defs.schema.json field-for-field.
"""
from .common import ClassificationLevelValue


class ClassificationLevel:
    """
    Sensitivity classification level for domain objects.
    
    Classification controls access and handling requirements:
    - public: No restrictions
    - internal: Internal use only
    - restricted: Need-to-know access
    - confidential: Strict access controls
    """
    
    def __init__(self, value: ClassificationLevelValue) -> None:
        self._value = value
    
    @classmethod
    def public(cls) -> "ClassificationLevel":
        return cls(ClassificationLevelValue.PUBLIC)
    
    @classmethod
    def internal(cls) -> "ClassificationLevel":
        return cls(ClassificationLevelValue.INTERNAL)
    
    @classmethod
    def restricted(cls) -> "ClassificationLevel":
        return cls(ClassificationLevelValue.RESTRICTED)
    
    @classmethod
    def confidential(cls) -> "ClassificationLevel":
        return cls(ClassificationLevelValue.CONFIDENTIAL)
    
    @property
    def value(self) -> ClassificationLevelValue:
        return self._value
    
    def __eq__(self, other: object) -> bool:
        if not isinstance(other, ClassificationLevel):
            return NotImplemented
        return self._value == other._value
    
    def __hash__(self) -> int:
        return hash(self._value)
    
    def __repr__(self) -> str:
        return f"ClassificationLevel({self._value.value})"
    
    def requires_access_control(self) -> bool:
        """Returns True if this classification requires access control."""
        return self._value != ClassificationLevelValue.PUBLIC


__all__ = ["ClassificationLevel"]

File: domain/classification_level.ts
/**
 * domain/classification_level.ts
 *
 * ClassificationLevel is the pure domain model for the sensitivity
 * classification of any platform object. It governs handling and disclosure
 * rules but implements no enforcement itself (enforcement is an
 * infrastructure/security concern outside the domain layer).
 *
 * Domain Purity: standard runtime only, no external imports.
 */

import { ClassificationLevelValue } from "./common.js";

/** Total order of classification levels, from least to most sensitive. */
const ORDER: ReadonlyArray<ClassificationLevelValue> = [
  ClassificationLevelValue.Public,
  ClassificationLevelValue.Internal,
  ClassificationLevelValue.Restricted,
  ClassificationLevelValue.Confidential,
];

/**
 * ClassificationLevel is an immutable value object wrapping a
 * ClassificationLevelValue, with pure comparison helpers. It never reaches
 * outside the domain to enforce access — it only represents the concept.
 */
export class ClassificationLevel {
  private constructor(public readonly value: ClassificationLevelValue) {
    Object.freeze(this);
  }

  static readonly Public = new ClassificationLevel(ClassificationLevelValue.Public);
  static readonly Internal = new ClassificationLevel(ClassificationLevelValue.Internal);
  static readonly Restricted = new ClassificationLevel(ClassificationLevelValue.Restricted);
  static readonly Confidential = new ClassificationLevel(ClassificationLevelValue.Confidential);

  static of(value: ClassificationLevelValue): ClassificationLevel {
    switch (value) {
      case ClassificationLevelValue.Public:
        return ClassificationLevel.Public;
      case ClassificationLevelValue.Internal:
        return ClassificationLevel.Internal;
      case ClassificationLevelValue.Restricted:
        return ClassificationLevel.Restricted;
      case ClassificationLevelValue.Confidential:
        return ClassificationLevel.Confidential;
      default:
        throw new Error(`Unknown ClassificationLevel value: ${value satisfies never}`);
    }
  }

  /** Rank of this level in the total order (0 = least sensitive). */
  rank(): number {
    return ORDER.indexOf(this.value);
  }

  /** True if this level is at least as sensitive as `other`. */
  atLeast(other: ClassificationLevel): boolean {
    return this.rank() >= other.rank();
  }

  /** Returns the more sensitive of two classification levels. */
  static max(a: ClassificationLevel, b: ClassificationLevel): ClassificationLevel {
    return a.rank() >= b.rank() ? a : b;
  }

  toString(): string {
    return this.value;
  }
}
